src/api/shop.py

```diff
@@ -0,0 +1,150 @@
+        treats = (
+            connection.execute(
+                sqlalchemy.text(
+                    """
+                    SELECT sku, name, price, satiety 
+                    FROM treats
+                    """
+                )
+            )
+            .mappings()
+            .fetchall()
+        )
+
+    return treats
+
+
+class Purchase(BaseModel):
+    user_id: int
+    order_id: int
+    treat_sku: str
+    quantity: int
+
+
+@router.post("/purchase")
+def purchase(purchase: Purchase):
+    """
+    Handles user purchasing a treat from the cafe.
+    """
+
+    if purchase.order_id < 0:
+        raise HTTPException(status_code=422, detail="order_id must be positive")
+
+    if purchase.quantity < 1:
+        raise HTTPException(status_code=422, detail="quantity must be at least 1")
+
+    try:
+        with db.engine.begin() as connection:
+            request = (
+                connection.execute(
+                    sqlalchemy.text(
+                        """
+                        SELECT
+                            u.id,
+                            g.gold,
+                            t.sku,
+                            t.price
+                        FROM users u
+                        JOIN gold_view g ON g.user_id = u.id
+                        CROSS JOIN treats t
+                        WHERE u.id = :user_id
+                            AND t.sku = :treat_sku
+                        """
+                    ),
+                    {"user_id": purchase.user_id, "treat_sku": purchase.treat_sku},
+                )
+                .mappings()
+                .fetchone()
+            )
+
+            if request is None:
+                raise HTTPException(
+                    status_code=404,
+                    detail=f"User {purchase.user_id} and/or treat {purchase.treat_sku} does not exist.",
+                )
+
+            cost = request["price"] * purchase.quantity
+
+            if cost > request["gold"]:
+                raise HTTPException(
+                    status_code=403,
+                    detail=f"The cost of the treats, {cost} gold, is greater than the user's total gold.",
+                )
+
+            success = (
+                connection.execute(
+                    sqlalchemy.text(
+                        """
+                        INSERT INTO purchases (order_id, user_id, item_sku, quantity)            
+                        VALUES (:order_id, :user_id, :treat_sku, :quantity)
+                        ON CONFLICT (order_id) DO NOTHING
+                        RETURNING order_id
+                        """
+                    ),
+                    {
+                        "order_id": purchase.order_id,
+                        "user_id": purchase.user_id,
+                        "treat_sku": purchase.treat_sku,
+                        "quantity": purchase.quantity,
+                    },
+                )
+                .mappings()
```

Comment:
`success` isn't used for anything but checking for existence, so you can probably leave out the `.mappings().fetchone()` entirely